src/Nixfmt/Pretty.hs

```diff
@@ -0,0 +1,141 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Nixfmt.Pretty where
+
+import           Data.List     hiding (break, group)
+import           Data.Text     (Text, pack)
+import qualified Data.Text     as Text
+import           Nixfmt.Predoc
+import           Nixfmt.Types
+import           Prelude       hiding (break)
+
+instance Pretty Trivium where
+    pretty EmptyLine            = emptyline
+    pretty (LineComment lc)     = text "#" <> pretty lc <> hardline
+    pretty (BlockComment bc)
+        = text "/*" <>
+          hcat (intersperse hardline $ map text bc) <>
+          text "*/" <> hardline
+
+instance Pretty [Trivium] where
+    pretty []     = mempty
+    pretty trivia = hardline <> hcat (map pretty trivia)
+
+instance Pretty NixToken where
+    pretty (Identifier i) = text i
+    pretty (EnvPath p)    = text ("<" <> p <> ">")
+    pretty (NixURI u)     = text u
+    pretty (NixFloat f)   = text f
+    pretty (NixInt i)     = text $ pack $ show i
+    pretty (NixText t)    = text t
+
+    pretty t              = pretty $ show t
```

Comment:
That’s quite error-prone. You will add new constructors to `NixToken` and will definitely forget to add cases to this definition, and the compiler will not warn you because of the catch-all pattern.